src/gui/qgsstackedwidget.h

```diff
@@ -0,0 +1,53 @@
+/***************************************************************************
+    qgsstackedwidget.h
+    ------------------
+    begin                : January 2024
+    copyright            : (C) 2024 by Stefanos Natsis
+    email                : uclaros at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSSTACKEDWIDGET_H
+#define QGSSTACKEDWIDGET_H
+
+#define SIP_NO_FILE
+
+#include <QStackedWidget>
+#include "qgis_gui.h"
+
+class QSize;
+
+/**
+ * \class QgsStackedWidget
+ * \ingroup gui
+ * \brief A QStackedWidget that can be shrunk to its current widget's size.
+ *
+ * A regular QStackedWidget can be shrunk down the size of its
+ * largest page widget. A QgsStackedWidget only takes the current
+ * page widget into account when resizing.
+ *
+ * \since QGIS 3.36
+ */
+class GUI_EXPORT QgsStackedWidget : public QStackedWidget
```

Comment:
Nice approach!

Can I suggest for future proofness that we:

1. Add an enum:
```
enum class SizeMode {
    ConsiderAllPages, //!< The sizes of all pages are considered when calculating the stacked widget size
    CurrentPage, //!< Only the size of the current page is considered when calculating the stacked widget size
};
```

2. Add getters/setters for SizeMode, default SizeMode to ConsiderAllPages (ie match QStackedWidget)
3. Adapt the subclass sizeHint methods to consider the SizeMode

This way we could start moving everything over from QStackedWidget to QgsStackedWidget, and we'd have a natural place to add other tweaks and improvements in future. Much like what we do for QgsScrollArea. (There's also a test to force use of QgsScrollArea over QScrollArea, we could potentially add the same for QgsStackedWidget later)

Reply:
Hmm, I guess we could, I'm not sure about the potential gain (other than homogeneity). In most places we use a stacked widget it either holds similar widgets or it is used as a temporary container and the inactive page gets deleted.
With QgsScrollArea it was quite different, as it implemented the absolutely necessary functionality to avoid messing with widgets while scrolling.

Reply:
There's definitely value in this minimal height stacked widget elsewhere. There's lots of places where we have ugly empty space in dialogs because a stacked widget has forced a too large minimum height for the current page, and all the widgets under that get pushed down for no visible reason.

I'd still push for the enum setting added, because this is kind of our one change to do that and have the default behavior match QStackedWidget. If we need to add other tweaks to QgsStackedWidget in future then things get a bit messier as we'd need to avoid breaking the initial default behaviour.